lib/client/session_windows.go

```diff
@@ -20,23 +20,208 @@ package client
+
+	// nodeClient is the parent of this session: the client connected to an
+	// SSH node
+	nodeClient *NodeClient
+
+	// Standard input/outputs for this session
+	stdin  io.Reader
+	stdout io.Writer
+	stderr io.Writer
+
+	// closer is used to simultaneously close all goroutines created by
+	// this session. It's also used to wait for everyone to close
+	closer *utils.CloseBroadcaster
+
 	ExitMsg string
+
+	enableEscapeSequences bool
+
+	conpty *conpty.ConPty
+}
+
+// requiredSyscalls lists all syscalls required to initialize the ConPTY.
+func requiredSyscalls() []string {
+	return []string{
+		"CreatePseudoConsole",
+		"ResizePseudoConsole",
+		"ClosePseudoConsole",
+		"InitializeProcThreadAttributeList",
+		"UpdateProcThreadAttribute",
+		"DeleteProcThreadAttributeList",
+	}
+}
+
+// ensureSyscallsExist validates that all listed Windows syscalls are present.
+func ensureSyscallsExist() error {
+	var missing []string
+
+	kernel32 := windows.NewLazySystemDLL("kernel32.dll")
+	for _, name := range requiredSyscalls() {
+		err := kernel32.NewProc(name).Find()
+		if err != nil {
+			missing = append(missing, name)
+		}
+	}
+
+	if len(missing) > 0 {
+		return fmt.Errorf(`syscalls required for tsh.exe ssh on Windows could not be found.
+Ensure Windows 10 1809 or later is installed.
+
+Missing syscalls: %s`, missing)
+	} else {
+		return nil
+	}
 }
 
-func newSession(
-	client *NodeClient,
+// initTerminal configures the terminal for raw, VT compatible input and output.
+// The returned function should be called before program exit to ensure the
+// terminal is reset, otherwise it will be left in a broken state.
+func initTerminal() (func(), error) {
+	stdoutFd := int(syscall.Stdout)
+	stdinFd := int(syscall.Stdin)
+
+	oldOutMode, err := winterm.GetConsoleMode(uintptr(stdoutFd))
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to retrieve stdout mode: %w", err)
+	}
+
+	oldInMode, err := winterm.GetConsoleMode(uintptr(stdinFd))
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to retrieve stdout mode: %w", err)
+	}
+
+	newOutMode := oldOutMode | winterm.ENABLE_VIRTUAL_TERMINAL_PROCESSING | winterm.DISABLE_NEWLINE_AUTO_RETURN
+
+	err = winterm.SetConsoleMode(uintptr(stdoutFd), newOutMode)
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to set stdout mode: %w", err)
+	}
+
+	newInMode := oldInMode
+	newInMode &^= winterm.ENABLE_ECHO_INPUT
+	newInMode &^= winterm.ENABLE_LINE_INPUT
+	newInMode &^= winterm.ENABLE_MOUSE_INPUT
+	newInMode &^= winterm.ENABLE_WINDOW_INPUT
+	newInMode &^= winterm.ENABLE_PROCESSED_INPUT
+
+	newInMode |= winterm.ENABLE_EXTENDED_FLAGS
+	newInMode |= winterm.ENABLE_INSERT_MODE
+	newInMode |= winterm.ENABLE_QUICK_EDIT_MODE
+	newInMode |= winterm.ENABLE_VIRTUAL_TERMINAL_INPUT
+
+	err = winterm.SetConsoleMode(uintptr(stdinFd), newInMode)
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to set stdin mode: %w", err)
+	}
+
+	return func() {
+		err = winterm.SetConsoleMode(uintptr(stdoutFd), oldOutMode)
```

Comment:
Do you intend for this to close over `err`?

Reply:
Whoops, nope!